File: app/src/main/java/dev/favourdevlabs/cleanthes/ui/home/HomeViewModel.java
package dev.favourdevlabs.cleanthes.ui.home;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import dev.favourdevlabs.cleanthes.data.entities.VaultEntry;
import dev.favourdevlabs.cleanthes.data.repository.VaultRepository;
import dev.favourdevlabs.cleanthes.ui.auth.SessionManager;

import java.util.List;
import java.util.ArrayList;
import javax.crypto.SecretKey;

public class HomeViewModel extends AndroidViewModel {

    private final MutableLiveData<List<VaultEntry>> filteredEntries = new MutableLiveData<>();

    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();

    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);

    private final MutableLiveData<List<String>> categories = new MutableLiveData<>();

    private final MutableLiveData<Integer> entryCount = new MutableLiveData<>(0);

    private List<VaultEntry> masterList = new ArrayList<>();

    private String currentQuery = "";

    private String currentCategory = "All";

    private final VaultRepository repository;

    public HomeViewModel(@NonNull Application application) {
        super(application);
        repository = VaultRepository.getInstance(application);
    }

    public LiveData<List<VaultEntry>> getFilteredEntries() {
        return filteredEntries;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<List<String>> getCategories() {
        return categories;
    }

    public LiveData<Integer> getEntryCount() {
        return entryCount;
    }

    public void loadEntries() {
        SecretKey key = SessionManager.getSessionKey();
        if (key == null) {
            errorMessage.postValue("Session expired. Please Unlock again");
            return;
        }

        isLoading.postValue(true);

        new Thread(() -> {
            try {
                List<VaultEntry> all = repository.getAllEntries(key);
                masterList = all;

                entryCount.postValue(all.size());

                List<String> cats = repository.getAllCategories();
                categories.postValue(cats);

                applyFilter();

            } catch (Exception e) {
                errorMessage.postValue("Failed to load entries: " + e.getMessage());
            } finally {
                isLoading.postValue(false);
            }
        }).start();
    }

    public void setSearchQuery(String query) {
        this.currentQuery = (query == null) ? "" : query.trim();
        applyFilter();
    }

    public void setCategory(String category) {
        this.currentCategory = category;
        applyFilter();
    }

    public void deleteEntry(long entryId) {
        new Thread(() -> {
            try {
                repository.deleteEntry(entryId);
                loadEntries();
            } catch (Exception e) {
                errorMessage.postValue("Failed to delete entry.");
            }
        }).start();
    }

    public void toggleFavorite(VaultEntry entry, String plainPassword) {
        SecretKey key = SessionManager.getSessionKey();
        if (key == null)
            return;

        entry.setFavorite(!entry.isFavorite());

        new Thread(() -> {
            try {
                repository.updateEntry(entry, plainPassword, key);
                loadEntries();
            } catch (Exception e) {
                errorMessage.postValue("Failed to update entry");
            }
        }).start();
    }

    private void applyFilter() {
        List<VaultEntry> result = new ArrayList<>();

        for (VaultEntry entry : masterList) {
            boolean matchesCategory = currentCategory.equals("All")
                    || entry.getCategory().equalsIgnoreCase(currentCategory);

            boolean matchesQuery = currentQuery.isEmpty()
                    || entry.getTitle().toLowerCase().contains(currentQuery.toLowerCase())
                    || entry.getUsername().toLowerCase().contains(currentQuery.toLowerCase());

            if (matchesCategory && matchesQuery) {
                result.add(entry);
            }
        }

        filteredEntries.postValue(result);
    }
}

File: app/src/main/java/dev/favourdevlabs/cleanthes/ui/home/HomeViewModel.kt
package dev.favourdevlabs.cleanthes.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import dev.favourdevlabs.cleanthes.data.entities.VaultEntry
import dev.favourdevlabs.cleanthes.ui.auth.SessionManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

import dev.favourdevlabs.cleanthes.domain.usecase.DeleteVaultEntryUseCase
import dev.favourdevlabs.cleanthes.domain.usecase.GetVaultEntriesUseCase
import dev.favourdevlabs.cleanthes.domain.usecase.SaveVaultEntryUseCase

data class HomeUiState(
    val isLoading: Boolean = false,
    val entries: List<VaultEntry> = emptyList(),
    val categories: List<String> = emptyList(),
    val entryCount: Int = 0,
    val searchQuery: String = "",
    val selectedCategory: String = "All",
    val pendingDeleteIds: Set<Long> = emptySet(),
    val errorMessage: String? = null,
) {
    val filteredEntries: List<VaultEntry>
        get() = entries
            .filter { it.id !in pendingDeleteIds }
            .filter { entry ->
                (selectedCategory == "All" ||
                        entry.category.equals(selectedCategory, ignoreCase = true)) &&
                (searchQuery.isEmpty() ||
                        entry.title.lowercase().contains(searchQuery.lowercase()) ||
                        (entry.username?.lowercase()?.contains(searchQuery.lowercase()) == true))
            }
}

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val getVaultEntries: GetVaultEntriesUseCase,
    private val deleteVaultEntry: DeleteVaultEntryUseCase,
    private val saveVaultEntry: SaveVaultEntryUseCase,
    private val sessionManager: SessionManager,
) : ViewModel() {

    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

   fun loadEntries() {
        val key = sessionManager.getSessionKey() ?: run {
            _uiState.update { it.copy(errorMessage = "Session expired. Please unlock again.") }
            return
        }
        _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            try {
                val result = withContext(Dispatchers.IO) { getVaultEntries(key) }
                _uiState.update {
                    it.copy(
                        isLoading  = false,
                        entries    = result.entries,
                        categories = result.categories,
                        entryCount = result.entries.size,
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isLoading = false, errorMessage = "Failed to load entries: ${e.message}")
                }
            }
        }
    } 

    fun setSearchQuery(query: String) =
        _uiState.update { it.copy(searchQuery = query.trim()) }

    fun setCategory(category: String) =
        _uiState.update { it.copy(selectedCategory = category) }

    fun onEntrySwipedToDelete(entryId: Long) =
        _uiState.update { it.copy(pendingDeleteIds = it.pendingDeleteIds + entryId) }

    fun undoDelete(entryId: Long) =
        _uiState.update { it.copy(pendingDeleteIds = it.pendingDeleteIds - entryId) }

   fun confirmDelete(entryId: Long) {
        _uiState.update { it.copy(pendingDeleteIds = it.pendingDeleteIds - entryId) }
        viewModelScope.launch {
            try {
                withContext(Dispatchers.IO) { deleteVaultEntry(entryId) }
                loadEntries()
            } catch (_: Exception) {
                _uiState.update { it.copy(errorMessage = "Failed to delete entry.") }
            }
        }
    } 

    fun clearError() = _uiState.update { it.copy(errorMessage = null) }

   fun toggleFavorite(entry: VaultEntry, plainPassword: String) {
        val key = sessionManager.getSessionKey() ?: return
        entry.isFavorite = !entry.isFavorite
        viewModelScope.launch {
            try {
                withContext(Dispatchers.IO) {
                    saveVaultEntry(SaveVaultEntryUseCase.Params.Edit(entry, plainPassword, key))
                }
                loadEntries()
            } catch (_: Exception) {
                _uiState.update { it.copy(errorMessage = "Failed to update entry.") }
            }
        }
    } 
}
